### ISSUE TITLE
Arithmetic expression

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#187]
Arithmetic expressions start with `.` and use infix notation, with usual precedence rules. Function calls use the mathematical `f(a, b)` notation instead of the shell-ish `(f a b)` notation. Barewords that look like variable names are treated as variables.

For instance, `. 1+p/q-sqrt(2)` is equivalent to `- (+ 1 (/ $p $q)) (sqrt 2)` (assuming #186 has been fixed).

Arithmetic expressions require a special parsing rule.

The unusual choice of `.` results from the author's strong detestation towards the clumsy `$(( ))` syntax used in bash and zsh, which discourages any use of the feature.

**Comment by krader1961** [comment]
I like this but would suggest using `math` instead of `.` for the command name. That is also consistent with the recently introduced math name space so you would write `math 1 + p / q - math:sqrt(2)`. The interface should allow multiple args without needing to quote the expression. It should implicitly join multiple args into a space separated string of tokens before parsing the expression.

One wrinkle involves disambiguating the magic floating point literals `inf` and `nan` if there are vars of the same name. One option is to introduce `math:inf` and `math:nan` vars and treat `inf` and `nan` barewords as user var references. Another option is to simply ignore vars of that name and treat the literals as the corresponding floating point values. I prefer the first option; although, that is inconsistent with how `inf` and `nan` are currently interpreted in other numeric contexts.

Presumably builtins like `count` will be usable so you can write an expression like `math 1 + count(arr)`; where `arr` is a var whose value is a list. What about arbitrary elvish functions that output a value which makes sense given the context. For example,

```elvish
fn my_rand {
    put 42
}

x = (math 1 + my_rand())
```

**Comment by xiaq** [comment]
This issue is low-priority and I'm not sure whether it's a good idea after all, but it doesn't hurt to talk about the design:

* I now feel that we should just require variable names to have `$`, to avoid the headaches and special cases you have pointed out. So `inf` will be infinity, while `$inf` will be the variable. Function names are distinguished by the `(` that follows them. So my original example `1+p/q-sqrt(2)` now reads `1+$p/$q-sqrt(2)`. It is slightly more verbose, but Perl manages fine.

* As for the choice of how to introduce arithmetic expression, I still prefer some kind of punctuation (doesn't have to be `.` though) instead of `math`, because this requires special parsing rules. Using `math` creates the delusion that it is a normal command, while punctuation hints at special syntax.

**Comment by zzamboni** [comment]
I think it would be interesting, but I'm not sure what a strong motivation would be for this. I personally find suffix notation much clearer and unambiguous. The moment you introduce infix notation you have to worry about precedence rules, for example, which are implicit in suffix.

**Comment by krader1961** [comment]
The current approach for math expressions is fine for people, like me, who prefer HP RPN calculators or are familiar with languages like Lisp. Infix math expressions will feel more natural to people who prefer TI calculators. Which is to say, most people. Whether we should cater to that group is an open question :smile_cat:

**Comment by xiaq** [comment]
It has never bothered me enough either, hence the `maybe` tag. I'm not in a hurry to implement anything but it's fun to think about the design.

**Comment by krader1961** [comment]
@xiaq, Good point about using punctuation rather than a word like `math` due to the need to invoke different parsing rules. Perhaps a single `=` would work well since that's what people who use TI style calculators are accustomed to pressing to get a result. Yes, that is potentially confusing since some people used to the POSIX `test` command might expect it to introduce an equality test but I think the risk is worth the improved legibility.

**Comment by hanche** [comment]
Since a dollar sign followed by a space is currently meaningless, it is also available for this purpose. It would make sense for a TeX user like me. :-) `</bikeshed>`

**Comment by hanche** [comment]
One more thought occurred to me: Users coming from a different shell might reasonably expect `.` to be used for sourcing a file. So I think it's better to leave it undefined. POLA and all that.

**Comment by xiaq** [comment]
Revisiting the choice of the punctuation, `=` would be a good choice. It eludes to the idea of doing arithmetics, and is already a conditional metacharacter.

**Comment by whoizit** [comment]
Polish notation is the only thing that scares me in Elvish. It's funny when expressions are simple, but when you have a complex formula, then writing in Polish notation without intermediate variables is a problem. This is a very good suggestion.
